File: ts/src/integrations.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { createRealm } from "./realm.js";
import { RealmError } from "./errors.js";

interface Captured {
  method: string;
  url: string;
  body?: unknown;
  authorization?: string;
}

/**
 * mkFetch distinguishes the platform bootstrap (grant_type=platform_api_key,
 * auto-issued to any /auth/login) from the mint (grant_type=
 * integration_installation), so the mint test can assert the mint response
 * rather than the bootstrap one.
 */
function mkFetch(handler: (req: Captured) => Response): typeof fetch {
  return (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const method = init?.method ?? "GET";
    let body: unknown;
    if (typeof init?.body === "string" && init.body.length > 0) {
      try { body = JSON.parse(init.body); } catch { body = init.body; }
    }
    const auth = init?.headers
      ? (init.headers as Record<string, string>)["authorization"]
      : undefined;
    if (url.endsWith("/auth/login") && (body as { grant_type?: string })?.grant_type === "platform_api_key") {
      return new Response(JSON.stringify({
        status: "ok", subject_type: "platform", refresh_token: "rtok-platform",
        access_token: "pt_x", expires_in: 300,
      }), { status: 200, headers: { "content-type": "application/json" } });
    }
    return handler({ method, url, body, authorization: auth });
  }) as typeof fetch;
}

const cfg = { realmId: "r", apiKey: "rk_live_x", baseUrl: "https://auth.test" };

test("integrations.register: POSTs to the platform route + maps camelCase", async () => {
  const fetch = mkFetch((req) => {
    assert.equal(req.method, "POST");
    assert.match(req.url, /\/platforms\/r\/integrations$/);
    assert.deepEqual(req.body, { slug: "hiring-motion", display_name: "Hiring Motion" });
    assert.equal(req.authorization, "Bearer pt_x");
    return new Response(JSON.stringify({
      id: "intg-1", realm_id: "r", slug: "hiring-motion", display_name: "Hiring Motion",
      listed: false, disabled: false,
    }), { status: 201, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ ...cfg, fetch });
  const out = await realm.integrations.register({ slug: "hiring-motion", displayName: "Hiring Motion" });
  assert.equal(out.id, "intg-1");
  assert.equal(out.slug, "hiring-motion");
});

/**
 * The install body must carry the ADR-101 D7 STATED PERMISSION LIST.
 *
 * This test previously asserted the body was `{integration_id, role_id}`,
 * which is why the SDK shipped broken against the live issuer for as long as
 * it did: the issuer replaced `role_id` with `permissions` and answers
 * `400 permissions_required`, while the test pinned the old shape and stayed
 * green. A test that asserts the implementation protects the bug.
 */
test("integrations.install: POSTs the permission list, and no role_id, to the tenant installations route", async () => {
  let seenBody: Record<string, unknown> = {};
  const fetch = mkFetch((req) => {
    assert.match(req.url, /\/tenants\/t1\/integration-installations$/);
    seenBody = req.body as Record<string, unknown>;
    return new Response(JSON.stringify({
      id: "inst-1", integration_id: "intg-1", permissions: ["users:read"],
      principal_user_id: "u-9", status: "installed",
    }), { status: 201, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ ...cfg, fetch });
  const out = await realm.integrations.install("t1", {
    integrationId: "intg-1",
    permissions: ["users:read"],
  });
  // Assert the whole body: checking only that `permissions` is present would
  // still pass if the client also sent the retired `role_id`.
  assert.deepEqual(seenBody, { integration_id: "intg-1", permissions: ["users:read"] });
  assert.equal("role_id" in seenBody, false);
  assert.equal(out.id, "inst-1");
  assert.equal(out.status, "installed");
  assert.deepEqual(out.permissions, ["users:read"]);
});

test("integrations.install: the three ADR-101 permission refusals surface on RealmError.code", async () => {
  for (const [code, status] of [
    ["permissions_required", 400],
    ["unknown_permission", 400],
    ["permissions_exceed_grantor", 403],
  ] as const) {
    const fetch = mkFetch(() =>
      new Response(JSON.stringify({ error: "no", code }),
        { status, headers: { "content-type": "application/json" } }));
    const realm = createRealm({ ...cfg, fetch });
    await assert.rejects(
      () => realm.integrations.install("t1", { integrationId: "intg-1", permissions: ["users:read"] }),
      (e: unknown) => e instanceof RealmError && e.code === code,
      `expected ${code} to reach RealmError.code`,
    );
  }
});

/**
 * RETAINED FOR THE MAPPING ONLY: the issuer has not emitted
 * `role_not_service_typed` since ADR-101 D7, so this asserts the code still
 * resolves for anyone branching on it, NOT that the refusal can still occur.
 */
test("integrations.install: role_not_service_typed still resolves (dead code, kept for compat)", async () => {
  const fetch = mkFetch(() =>
    new Response(JSON.stringify({ error: "no", code: "role_not_service_typed" }),
      { status: 400, headers: { "content-type": "application/json" } }));
  const realm = createRealm({ ...cfg, fetch });
  await assert.rejects(
    () => realm.integrations.install("t1", { integrationId: "intg-1", permissions: ["users:read"] }),
    (e: unknown) => e instanceof RealmError && e.code === "role_not_service_typed",
  );
});

test("integrations.listInstallations: decodes the inbound-access page", async () => {
  const fetch = mkFetch(() =>
    new Response(JSON.stringify({
      items: [{ id: "inst-1", integration_id: "intg-1", permissions: ["users:read", "users:manage"], mint_count: 3 }],
      next_cursor: null,
    }), { status: 200, headers: { "content-type": "application/json" } }));
  const realm = createRealm({ ...cfg, fetch });
  const page = await realm.integrations.listInstallations("t1");
  assert.equal(page.items.length, 1);
  assert.deepEqual(page.items[0]?.permissions, ["users:read", "users:manage"]);
  assert.equal(page.items[0]?.mint_count, 3);
  assert.equal(page.next_cursor, null);
});

test("integrations.uninstall: DELETEs the installation", async () => {
  let seen = "";
  const fetch = mkFetch((req) => {
    seen = `${req.method} ${new URL(req.url).pathname}`;
    return new Response(JSON.stringify({ status: "ok" }), { status: 200, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ ...cfg, fetch });
  await realm.integrations.uninstall("t1", "inst-1");
  assert.equal(seen, "DELETE /tenants/t1/integration-installations/inst-1");
});

test("integrations.mintToken: sends the raw api key with NO bearer + decodes access-only token", async () => {
  const fetch = mkFetch((req) => {
    // The mint hits /auth/login with grant_type=integration_installation.
    assert.match(req.url, /\/auth\/login$/);
    assert.deepEqual(req.body, {
      grant_type: "integration_installation",
      api_key: "rk_live_src",
      installation_id: "inst-1",
      source_org_id: "org-a",
    });
    // The raw key is the credential — it must NOT ride as a bearer.
    assert.equal(req.authorization, undefined);
    return new Response(JSON.stringify({
      status: "ok", subject_type: "service",
      access_token: "brokered-jwt", expires_in: 600, tenant_id: "t-target", role: "svc",
    }), { status: 200, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ ...cfg, fetch });
  const out = await realm.integrations.mintToken({
    apiKey: "rk_live_src", installationId: "inst-1", sourceOrgId: "org-a",
  });
  assert.equal(out.access_token, "brokered-jwt");
  assert.equal(out.expires_in, 600);
  assert.equal(out.role, "svc");
  // No refresh token in the shape — access-only (ADR-083 §7.5).
  assert.equal((out as Record<string, unknown>)["refresh_token"], undefined);
});

test("integrations.mintToken: key_class_mismatch surfaces on RealmError.code", async () => {
  const fetch = mkFetch(() =>
    new Response(JSON.stringify({ error: "no", code: "key_class_mismatch" }),
      { status: 401, headers: { "content-type": "application/json" } }));
  const realm = createRealm({ ...cfg, fetch });
  await assert.rejects(
    () => realm.integrations.mintToken({ apiKey: "rk_live_svc", installationId: "inst-1", sourceOrgId: "o" }),
    (e: unknown) => e instanceof RealmError && e.code === "key_class_mismatch",
  );
});

test("integrations.disable/enable/remove: hit the source lifecycle verbs", async () => {
  const hits: string[] = [];
  const fetch = mkFetch((req) => {
    hits.push(`${req.method} ${new URL(req.url).pathname}`);
    return new Response(JSON.stringify({ status: "ok" }), { status: 200, headers: { "content-type": "application/json" } });
  });
  const realm = createRealm({ ...cfg, fetch });
  await realm.integrations.disable("intg-1");
  await realm.integrations.enable("intg-1");
  await realm.integrations.remove("intg-1");
  assert.deepEqual(hits, [
    "POST /platforms/r/integrations/intg-1/disable",
    "POST /platforms/r/integrations/intg-1/enable",
    "DELETE /platforms/r/integrations/intg-1",
  ]);
});

File: go/integrations_test.go
package realmid

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"testing"
)

func newIntegrationsRealm(t *testing.T, url string) *Realm {
	t.Helper()
	r, _ := NewRealm(Config{RealmID: testRealmID, APIKey: "rk", BaseURL: url})
	return r
}

// TestIntegrations_Register asserts the source-side register posts to the
// platform integrations route and decodes the result.
func TestIntegrations_Register(t *testing.T) {
	var gotPath string
	var gotBody map[string]any
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/platforms/" + testRealmID + "/integrations": func(w http.ResponseWriter, r *http.Request) {
			gotPath = r.URL.Path
			buf, _ := io.ReadAll(r.Body)
			_ = json.Unmarshal(buf, &gotBody)
			w.WriteHeader(http.StatusCreated)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id": "intg-1", "realm_id": testRealmID, "slug": "hiring-motion",
				"display_name": "Hiring Motion", "listed": false, "disabled": false,
			})
		},
	})
	defer srv.Close()

	out, err := newIntegrationsRealm(t, srv.URL).Integrations.Register(context.Background(),
		IntegrationCreate{Slug: "hiring-motion", DisplayName: "Hiring Motion"})
	if err != nil {
		t.Fatalf("Register: %v", err)
	}
	if gotPath != "/platforms/"+testRealmID+"/integrations" {
		t.Errorf("path = %q", gotPath)
	}
	if gotBody["slug"] != "hiring-motion" {
		t.Errorf("body slug = %v", gotBody["slug"])
	}
	if out.ID != "intg-1" || out.Slug != "hiring-motion" {
		t.Errorf("decoded = %+v", out)
	}
}

// TestIntegrations_SlugTakenMapsSentinel: 409 slug_taken → ErrIntegrationSlugTaken.
func TestIntegrations_SlugTakenMapsSentinel(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/platforms/" + testRealmID + "/integrations": func(w http.ResponseWriter, _ *http.Request) {
			w.WriteHeader(http.StatusConflict)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "taken", "code": "slug_taken"})
		},
	})
	defer srv.Close()
	_, err := newIntegrationsRealm(t, srv.URL).Integrations.Register(context.Background(),
		IntegrationCreate{Slug: "dup", DisplayName: "Dup"})
	if !errors.Is(err, ErrIntegrationSlugTaken) {
		t.Errorf("want ErrIntegrationSlugTaken, got %v", err)
	}
}

// TestIntegrations_Install asserts the target-side install posts to the tenant
// installations route with the STATED PERMISSION LIST (ADR-101 D7).
//
// This test previously asserted the body carried `role_id`, which is why the
// SDK shipped broken against the live issuer for as long as it did: the issuer
// replaced role_id with `permissions` and answers `400 permissions_required`,
// while the test pinned the old shape and stayed green. A test that asserts
// the implementation protects the bug.
func TestIntegrations_Install(t *testing.T) {
	var gotBody map[string]any
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/tenants/t1/integration-installations": func(w http.ResponseWriter, r *http.Request) {
			buf, _ := io.ReadAll(r.Body)
			_ = json.Unmarshal(buf, &gotBody)
			w.WriteHeader(http.StatusCreated)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"id": "inst-1", "integration_id": "intg-1",
				"permissions":       []string{"users:read"},
				"principal_user_id": "u-9", "status": "installed",
			})
		},
	})
	defer srv.Close()
	out, err := newIntegrationsRealm(t, srv.URL).Integrations.Install(context.Background(), "t1",
		InstallRequest{IntegrationID: "intg-1", Permissions: []string{"users:read"}})
	if err != nil {
		t.Fatalf("Install: %v", err)
	}
	if gotBody["integration_id"] != "intg-1" {
		t.Errorf("integration_id: body = %v", gotBody)
	}
	// The wire field must be `permissions`, and `role_id` must be ABSENT —
	// asserting only the former would still pass if both were sent.
	perms, ok := gotBody["permissions"].([]any)
	if !ok || len(perms) != 1 || perms[0] != "users:read" {
		t.Errorf("permissions: body = %v", gotBody)
	}
	if _, present := gotBody["role_id"]; present {
		t.Errorf("role_id must not be sent — the issuer retired it (ADR-101 D7): %v", gotBody)
	}
	if out.ID != "inst-1" || out.Status != "installed" {
		t.Errorf("decoded = %+v", out)
	}
	if len(out.Permissions) != 1 || out.Permissions[0] != "users:read" {
		t.Errorf("decoded permissions = %+v", out.Permissions)
	}
}

// TestIntegrations_PermissionErrorsMapSentinels covers the three refusals the
// permission-stated install can produce. None of them existed while the SDK
// was still sending role_id, so none had a sentinel.
func TestIntegrations_PermissionErrorsMapSentinels(t *testing.T) {
	for _, tc := range []struct {
		code   string
		status int
		want   error
	}{
		{"permissions_required", http.StatusBadRequest, ErrPermissionsRequired},
		{"unknown_permission", http.StatusBadRequest, ErrUnknownPermission},
		{"permissions_exceed_grantor", http.StatusForbidden, ErrPermissionsExceedGrantor},
	} {
		t.Run(tc.code, func(t *testing.T) {
			srv := authTestServer(t, map[string]http.HandlerFunc{
				"/tenants/t1/integration-installations": func(w http.ResponseWriter, r *http.Request) {
					w.WriteHeader(tc.status)
					_ = json.NewEncoder(w).Encode(map[string]any{
						"error": map[string]any{"code": tc.code, "message": tc.code},
					})
				},
			})
			defer srv.Close()
			_, err := newIntegrationsRealm(t, srv.URL).Integrations.Install(context.Background(), "t1",
				InstallRequest{IntegrationID: "intg-1", Permissions: []string{"users:read"}})
			if !errors.Is(err, tc.want) {
				t.Fatalf("want %v, got %v", tc.want, err)
			}
		})
	}
}

// TestIntegrations_RoleNotServiceTypedMapsSentinel: 400 role_not_service_typed
// → ErrRoleNotServiceTyped. RETAINED FOR THE MAPPING ONLY: the issuer has not
// emitted this code since ADR-101 D7, so this asserts the sentinel still
// resolves for anyone matching it, NOT that the refusal can still occur.
func TestIntegrations_RoleNotServiceTypedMapsSentinel(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/tenants/t1/integration-installations": func(w http.ResponseWriter, _ *http.Request) {
			w.WriteHeader(http.StatusBadRequest)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "no", "code": "role_not_service_typed"})
		},
	})
	defer srv.Close()
	_, err := newIntegrationsRealm(t, srv.URL).Integrations.Install(context.Background(), "t1",
		InstallRequest{IntegrationID: "intg-1", Permissions: []string{"users:read"}})
	if !errors.Is(err, ErrRoleNotServiceTyped) {
		t.Errorf("want ErrRoleNotServiceTyped, got %v", err)
	}
}

// TestIntegrations_ListInstallations decodes the inbound-access page.
func TestIntegrations_ListInstallations(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/tenants/t1/integration-installations": func(w http.ResponseWriter, _ *http.Request) {
			_ = json.NewEncoder(w).Encode(map[string]any{
				"items": []any{
					map[string]any{"id": "inst-1", "integration_id": "intg-1",
						"permissions": []string{"users:read", "users:manage"}, "mint_count": 3},
				},
				"next_cursor": nil,
			})
		},
	})
	defer srv.Close()
	page, err := newIntegrationsRealm(t, srv.URL).Integrations.ListInstallations(context.Background(), "t1", nil)
	if err != nil {
		t.Fatalf("ListInstallations: %v", err)
	}
	if len(page.Items) != 1 || page.Items[0].MintCount != 3 ||
		len(page.Items[0].Permissions) != 2 || page.Items[0].Permissions[0] != "users:read" {
		t.Errorf("decoded = %+v", page.Items)
	}
}

// TestIntegrations_Uninstall asserts DELETE on the installation route.
func TestIntegrations_Uninstall(t *testing.T) {
	var gotMethod, gotPath string
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/tenants/t1/integration-installations/inst-1": func(w http.ResponseWriter, r *http.Request) {
			gotMethod, gotPath = r.Method, r.URL.Path
			_ = json.NewEncoder(w).Encode(map[string]any{"status": "ok"})
		},
	})
	defer srv.Close()
	if err := newIntegrationsRealm(t, srv.URL).Integrations.Uninstall(context.Background(), "t1", "inst-1"); err != nil {
		t.Fatalf("Uninstall: %v", err)
	}
	if gotMethod != "DELETE" || gotPath != "/tenants/t1/integration-installations/inst-1" {
		t.Errorf("method/path = %s %s", gotMethod, gotPath)
	}
}

// TestIntegrations_MintToken asserts the mint posts the integration_installation
// grant with the raw api key (NO Authorization header) and decodes the token.
func TestIntegrations_MintToken(t *testing.T) {
	var gotAuth string
	var gotBody map[string]any
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/login": func(w http.ResponseWriter, r *http.Request) {
			gotAuth = r.Header.Get("Authorization")
			buf, _ := io.ReadAll(r.Body)
			_ = json.Unmarshal(buf, &gotBody)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"status": "ok", "subject_type": "service",
				"access_token": "brokered-jwt", "expires_in": 600,
				"tenant_id": "t-target", "role": "svc",
			})
		},
	})
	defer srv.Close()
	out, err := newIntegrationsRealm(t, srv.URL).Integrations.MintToken(context.Background(),
		IntegrationMintRequest{APIKey: "rk_live_src", InstallationID: "inst-1", SourceOrgID: "org-a"})
	if err != nil {
		t.Fatalf("MintToken: %v", err)
	}
	if gotBody["grant_type"] != "integration_installation" {
		t.Errorf("grant_type = %v", gotBody["grant_type"])
	}
	if gotBody["api_key"] != "rk_live_src" || gotBody["source_org_id"] != "org-a" {
		t.Errorf("body = %v", gotBody)
	}
	// The raw key is the credential — it must NOT ride as a bearer.
	if gotAuth != "" {
		t.Errorf("mint must not send Authorization; got %q", gotAuth)
	}
	if out.AccessToken != "brokered-jwt" || out.ExpiresIn != 600 || out.Role != "svc" {
		t.Errorf("decoded = %+v", out)
	}
}

// TestIntegrations_MintKeyClassMismatchMapsSentinel: 401 key_class_mismatch →
// ErrKeyClassMismatch.
func TestIntegrations_MintKeyClassMismatchMapsSentinel(t *testing.T) {
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/auth/login": func(w http.ResponseWriter, _ *http.Request) {
			w.WriteHeader(http.StatusUnauthorized)
			_ = json.NewEncoder(w).Encode(map[string]any{"error": "no", "code": "key_class_mismatch"})
		},
	})
	defer srv.Close()
	_, err := newIntegrationsRealm(t, srv.URL).Integrations.MintToken(context.Background(),
		IntegrationMintRequest{APIKey: "rk_live_svc", InstallationID: "inst-1", SourceOrgID: "org-a"})
	if !errors.Is(err, ErrKeyClassMismatch) {
		t.Errorf("want ErrKeyClassMismatch, got %v", err)
	}
}

// TestIntegrations_DisableEnableRemove exercises the source lifecycle verbs.
func TestIntegrations_DisableEnableRemove(t *testing.T) {
	hits := map[string]string{}
	h := func(w http.ResponseWriter, r *http.Request) {
		hits[r.URL.Path] = r.Method
		_ = json.NewEncoder(w).Encode(map[string]any{"status": "ok"})
	}
	srv := authTestServer(t, map[string]http.HandlerFunc{
		"/platforms/" + testRealmID + "/integrations/intg-1/disable": h,
		"/platforms/" + testRealmID + "/integrations/intg-1/enable":  h,
		"/platforms/" + testRealmID + "/integrations/intg-1":         h,
	})
	defer srv.Close()
	c := newIntegrationsRealm(t, srv.URL).Integrations
	if err := c.Disable(context.Background(), "intg-1"); err != nil {
		t.Fatalf("Disable: %v", err)
	}
	if err := c.Enable(context.Background(), "intg-1"); err != nil {
		t.Fatalf("Enable: %v", err)
	}
	if err := c.Remove(context.Background(), "intg-1"); err != nil {
		t.Fatalf("Remove: %v", err)
	}
	if hits["/platforms/"+testRealmID+"/integrations/intg-1/disable"] != "POST" ||
		hits["/platforms/"+testRealmID+"/integrations/intg-1/enable"] != "POST" ||
		hits["/platforms/"+testRealmID+"/integrations/intg-1"] != "DELETE" {
		t.Errorf("hits = %v", hits)
	}
}
